sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMap.java

```diff
@@ -225,4 +232,93 @@ public void printPerfMetrics() {
   }
 
+  /**
+   * Sorts the key, value data in this map in place, and return them as an iterator.
+   *
+   * The only memory that is allocated is the address/prefix array, 16 bytes per record.
+   */
+  public KVIterator<UnsafeRow, UnsafeRow> sortedIterator() {
+    int numElements = map.numElements();
+    final int numKeyFields = groupingKeySchema.size();
+    TaskMemoryManager memoryManager = map.getTaskMemoryManager();
+
+    UnsafeExternalRowSorter.PrefixComputer prefixComp =
+      SortPrefixUtils.createPrefixGenerator(groupingKeySchema);
+    PrefixComparator prefixComparator = SortPrefixUtils.getPrefixComparator(groupingKeySchema);
+
+    final BaseOrdering ordering = GenerateOrdering.create(groupingKeySchema);
+    RecordComparator recordComparator = new RecordComparator() {
+      private final UnsafeRow row1 = new UnsafeRow();
+      private final UnsafeRow row2 = new UnsafeRow();
+
+      @Override
+      public int compare(Object baseObj1, long baseOff1, Object baseObj2, long baseOff2) {
+        row1.pointTo(baseObj1, baseOff1 + 4, numKeyFields, -1);
+        row2.pointTo(baseObj2, baseOff2 + 4, numKeyFields, -1);
+        return ordering.compare(row1, row2);
+      }
+    };
+
+    // Insert the records into the in-memory sorter.
+    final UnsafeInMemorySorter sorter = new UnsafeInMemorySorter(
+      memoryManager, recordComparator, prefixComparator, numElements);
+
+    BytesToBytesMap.BytesToBytesMapIterator iter = map.iterator();
+    UnsafeRow row = new UnsafeRow();
+    while (iter.hasNext()) {
+      final BytesToBytesMap.Location loc = iter.next();
+      final Object baseObject = loc.getKeyAddress().getBaseObject();
+      final long baseOffset = loc.getKeyAddress().getBaseOffset();
+
+      // Get encoded memory address
+      MemoryBlock page = loc.getMemoryPage();
+      long address = memoryManager.encodePageNumberAndOffset(page, baseOffset - 8);
+
+      // Compute prefix
+      row.pointTo(baseObject, baseOffset, numKeyFields, loc.getKeyLength());
+      final long prefix = prefixComp.computePrefix(row);
+
+      sorter.insertRecord(address, prefix);
+    }
+
+    // Return the sorted result as an iterator.
+    return new KVIterator<UnsafeRow, UnsafeRow>() {
+
+      private UnsafeSorterIterator sortedIterator = sorter.getSortedIterator();
+      private final UnsafeRow key = new UnsafeRow();
+      private final UnsafeRow value = new UnsafeRow();
+      private int numValueFields = aggregationBufferSchema.size();
+
+      @Override
+      public boolean next() throws IOException {
+        if (sortedIterator.hasNext()) {
+          sortedIterator.loadNext();
+          Object baseObj = sortedIterator.getBaseObject();
+          long recordOffset = sortedIterator.getBaseOffset();
+          int recordLen = sortedIterator.getRecordLength();
+          int keyLen = PlatformDependent.UNSAFE.getInt(baseObj, recordOffset);
```

Comment:
This points to the key length because `recordOffset` is 4 bytes past where the total record length was stored?